src/components/timeline/TrackScreenshots.js

```diff
@@ -200,24 +201,38 @@ class HoverPreview extends PureComponent<HoverPreviewProps> {
     const payload: ScreenshotPayload = (screenshots[screenshotIndex].data: any);
     const { url, windowWidth, windowHeight } = payload;
 
+    // Compute hover image size coefficients to ensure it does not exceed top of screen
+    // To prevent blurring, do not scale up the screenshots
+    const coefficient = Math.min(
+      1,
+      2 * (TRACK_HEIGHT / 2 + containerTop) / windowHeight,
+      MAXIMUM_HOVER_HEIGHT / windowHeight
+    );
     // Compute the hover image's thumbnail size.
-    let hoverHeight = HOVER_HEIGHT;
-    let hoverWidth = HOVER_HEIGHT / windowHeight * windowWidth;
+    let hoverHeight = windowHeight * coefficient;
+    let hoverWidth = windowWidth * coefficient;
 
-    if (hoverWidth > HOVER_HEIGHT * HOVER_MAX_WIDTH_RATIO) {
+    if (hoverWidth > hoverHeight * HOVER_MAX_WIDTH_RATIO) {
       // This is a really wide image, limit the height so it lays out reasonably.
-      hoverWidth = HOVER_HEIGHT * HOVER_MAX_WIDTH_RATIO;
+      hoverWidth = hoverHeight * HOVER_MAX_WIDTH_RATIO;
       hoverHeight = hoverWidth / windowWidth * windowHeight;
     }
 
     // Set the top so it centers around the track.
-    const top = containerTop + (TRACK_HEIGHT - hoverHeight) * 0.5;
-    const left =
-      offsetX + hoverWidth * 0.5 > width
-        ? // Stick the hover image on to the right side of the container.
-          pageX - offsetX + width - hoverWidth
-        : // Center the hover image around the mouse.
-          pageX - hoverWidth * 0.5;
+    let top = containerTop + (TRACK_HEIGHT - hoverHeight) * 0.5;
+    if (top < 0) {
+      // Stick the hover image on to the top side of the container
+      top = containerTop;
+    }
+    // Center the hover image around the mouse.
+    let left = pageX - hoverWidth * 0.5;
+    if (left < 0) {
+      // Stick the hover image on to the left side of the page.
+      left = 0;
+    } else if (offsetX + hoverWidth * 0.5 > width) {
+      // Stick the hover image on to the right side of the container.
+      left = pageX - offsetX + width - hoverWidth;
+    }
```

Comment:
It's making me think because `else if (offsetX + hoverWidth * 0.5 > width) {` is not using `left` at all and it's just assigning the `left` without using the first assigned value. There are a few better solutions that comes to my mind. First one is keeping the previous ternary operator and adding `if (left < 0) after the first assignment.
The second is doing something like that:
```js
if (offsetX + hoverWidth * 0.5 > width) {
  // Stick the hover image on to the right side of the container.
  left = pageX - offsetX + width - hoverWidth;
} else {
  // Center the hover image around the mouse.
  left = Math.min(0, pageX - hoverWidth * 0.5);
}
```
But I think first approach is a little bit clean. It's a bit personal preference though, maybe @julienw has better ideas?

Reply:
I wonder if the condition couldn't just be `else if (left + hoverWidth > width)`? This makes it super clear about what we're doing here.

Reply:
Doing ```else if (left + hoverWidth > width) ``` sticks left early due to offset. I made it ```else if (left + hoverWidth > width + (pageX - offsetX)) ``` I hope it makes more clear.